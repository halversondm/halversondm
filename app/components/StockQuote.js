/**
 * Created by Daniel on 6/26/2016.
 */
"use strict";

import React from "react";
import $ from "jQuery";

const StockQuote = React.createClass({
  getInitialState() {
    return {
      stockInput: "",
      stocks: []
    };
  },
  componentDidMount() {
    const stockList = ["MSFT", "AAPL", "JPM", "AMZN", "T", "F"];
    stockList.forEach(stock => {
      setTimeout(this.callService(stock), 1000);
    });
  },
  inputStock(event) {
    this.setState({stockInput: event.target.value});
  },
  submit() {
    this.callService(this.state.stockInput);
    this.setState({stockInput: ""});
  },
  callService(stockSymbol) {
    var url = "http://dev.markitondemand.com/Api/v2/Quote/jsonp?&symbol=" + stockSymbol;
    $.ajax({
      url: url,
      dataType: "jsonp"
    }).then(data => {
      var stocks = this.state.stocks;
      stocks.push(data);
      this.setState({stocks: stocks});
    }, () => {
      var stocks = this.state.stocks;
      var stock = {Symbol: stockSymbol, Name: "Not Found"};
      stocks.push(stock);
      this.setState({stocks: stocks});
    });
  },
  render() {
    return <div>
      <h2 className="text-primary">Stock Quotes</h2>
      <form className="form-inline">
        <div className="form-group">
          <label className="sr-only" htmlFor="symbol">Stock Symbol</label>
          <input type="text" name="stock" id="symbol"
                 className="form-control input-sm" placeholder="Symbol"
                 value={this.state.stockInput}
                 onChange={this.inputStock}/>
          <input type="button" className="btn-sm btn-default" value="Go!"
                 onClick={this.submit}/>
        </div>
      </form>
      <h4 className="text-primary">Stock List</h4>
      <table className="table table-striped">
        <thead>
        <tr>
          <th>Symbol</th>
          <th>Name</th>
          <th>Last Price</th>
          <th>Time</th>
          <th>Market Cap</th>
          <th>Change YTD</th>
          <th>High</th>
          <th>Low</th>
          <th>Open</th>
        </tr>
        </thead>
        <tbody>
        {
          this.state.stocks.map((stock, i) => {
            return <tr key={i}>
              <td>{stock.Symbol}</td>
              <td>{stock.Name}</td>
              <td>{stock.LastPrice}</td>
              <td>{stock.Timestamp}</td>
              <td>{stock.MarketCap}</td>
              <td>{stock.ChangeYTD}</td>
              <td>{stock.High}</td>
              <td>{stock.Low}</td>
              <td>{stock.Open}</td>
            </tr>;
          })
        }
        </tbody>
      </table>
    </div>;
  }
});

export default StockQuote;
