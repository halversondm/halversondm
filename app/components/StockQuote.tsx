/**
 * Created by Daniel on 6/26/2016.
 */
import * as React from 'react'
import { type ReactNode, useState, useEffect } from 'react'

interface StockQuoteState {
  stockInput: string
  stocks: Stock[]
}

interface Stock {
  Symbol: string
  Name: string
  LastPrice?: string
  Timestamp?: string
  MarketCap?: string
  ChangeYTD?: string
  High?: string
  Open?: string
  Low?: string
}

export default function StockQuote (): ReactNode {
  const stockList = ['MSFT', 'AAPL', 'JPM', 'AMZN', 'T', 'F']
  const [state, setState] = useState<StockQuoteState>(initialState())

  function initialState (): StockQuoteState {
    return {
      stockInput: '',
      stocks: []
    }
  }

  useEffect(() => {
    stockList.forEach((stock) => {
      callService(stock)
    })
  }, [])

  function inputStock (event): void {
    setState({ ...state, stockInput: event.target.value })
  }

  function submit (): void {
    callService(state.stockInput)
    setState({ ...state, stockInput: '' })
  }

  function callService (stockSymbol: string): void {
    const stocks = state.stocks
    const xhr = new XMLHttpRequest()
    const url = `/stock?stockSymbol=${stockSymbol}`
    xhr.open('POST', url)
    xhr.onload = () => {
      if (xhr.status >= 200 && xhr.status < 400) {
        stocks.push(JSON.parse(xhr.responseText))
      } else {
        console.log('unsucc ', xhr.responseText)
        const stock: Stock = { Symbol: stockSymbol, Name: 'Not Found' }
        stocks.push(stock)
      }
      return stocks
    }
    xhr.onerror = () => {
      console.log(xhr)
    }
    xhr.send()
  }

  return (
            <div>
                <h2 className="text-primary">Stock Quotes</h2>
                <form className="form-inline">
                    <div className="form-group">
                        <label className="sr-only" htmlFor="symbol">Stock Symbol</label>
                        <input type="text" name="stock" id="symbol"
                               className="form-control input-sm" placeholder="Symbol"
                               value={state.stockInput}
                               onChange={inputStock}/>
                        <input type="button" className="btn-sm btn-default" value="Go!"
                               onClick={submit}/>
                    </div>
                </form>
                <h4 className="text-primary">Stock List</h4>
                <table className="table table-striped">
                    <thead>
                    <tr>
                        <th>Symbol</th>
                        <th>Name</th>
                        <th>Last Price</th>
                        <th>Time</th>
                        <th>Market Cap</th>
                        <th>Change YTD</th>
                        <th>High</th>
                        <th>Low</th>
                        <th>Open</th>
                    </tr>
                    </thead>
                    <tbody>
                    {
                        state.stocks.map((stock, i) => {
                          return <tr key={i}>
                                <td>{stock.Symbol}</td>
                                <td>{stock.Name}</td>
                                <td>{stock.LastPrice}</td>
                                <td>{stock.Timestamp}</td>
                                <td>{stock.MarketCap}</td>
                                <td>{stock.ChangeYTD}</td>
                                <td>{stock.High}</td>
                                <td>{stock.Low}</td>
                                <td>{stock.Open}</td>
                            </tr>
                        })
                    }
                    </tbody>
                </table>
            </div>
  )
}
