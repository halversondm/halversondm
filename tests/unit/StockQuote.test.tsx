import React from "react";
import { render } from '@testing-library/react';
import StockQuote from "../../app/components/StockQuote";

test("StockQuote component", () => {
    const unit = <StockQuote />;
    const component = render(unit);
    const tree = component.container;
    expect(tree.toString()).toMatchSnapshot();
});
