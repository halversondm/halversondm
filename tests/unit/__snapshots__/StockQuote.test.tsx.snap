// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StockQuote component 1`] = `"<div><div><h2 class="text-primary">Stock Quotes</h2><form class="form-inline"><div class="form-group"><label class="sr-only" for="symbol">Stock Symbol</label><input id="symbol" class="form-control input-sm" placeholder="Symbol" type="text" value="" name="stock"><input class="btn-sm btn-default" type="button" value="Go!"></div></form><h4 class="text-primary">Stock List</h4><table class="table table-striped"><thead><tr><th>Symbol</th><th>Name</th><th>Last Price</th><th>Time</th><th>Market Cap</th><th>Change YTD</th><th>High</th><th>Low</th><th>Open</th></tr></thead><tbody></tbody></table></div></div>"`;
